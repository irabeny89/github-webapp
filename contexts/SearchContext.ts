import { createContext, Dispatch } from "react";
import { InitialSearchStateActionType, InitialSearchStateType } from "types";
import { initialState } from "components/SearchProvider";

const SearchContext = createContext<
  [InitialSearchStateType, Dispatch<InitialSearchStateActionType>]
>([initialState, () => {}]);

export default SearchContext;
