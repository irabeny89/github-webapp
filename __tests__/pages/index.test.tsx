import { initialState } from "@/contexts/SearchContext";
import Home from "@/pages/index";
import { render, screen, cleanup } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { rest } from "msw";
import { setupServer } from "msw/node";
import type { SearchDataType } from "types";

const searchTerm = "jest",
  page1Items = [
    {
      ...initialState.searchData,
      id: 1,
      owner: {
        id: 1,
        html_url: "https://example.com",
        login: "jester",
      },
      stargazers_count: 1,
      name: searchTerm,
      html_url: "https://example.com",
    },
  ],
  page2Items = [
    {
      ...initialState.searchData,
      id: 2,
      owner: {
        id: 2,
        html_url: "https://example.com",
        login: "jester2",
      },
      stargazers_count: 2,
      name: searchTerm + "2",
      html_url: "https://example.com",
    },
  ],
  page1response: SearchDataType = {
    ...initialState.searchData,
    items: page1Items,
  },
  page2response: SearchDataType = {
    ...initialState.searchData,
    items: page2Items,
  },
  testServer = setupServer(
    rest.get(`https://api.github.com/search/repositories`, (req, res, ctx) => {
      return req.headers.get("Accept")
        ? req.url.searchParams.get("page") === "1"
          ? res(ctx.json(page1response))
          : res(ctx.json(page2response))
        : res(ctx.status(400, "Invalid Accept header value"));
    })
  );

describe("Home page test", () => {
  beforeAll(() => testServer.listen());
  beforeEach(() => render(<Home />));
  afterEach(() => {
    testServer.resetHandlers(), cleanup;
  });
  afterAll(() => testServer.close());

  it("doesn't search when text box is empty.", async () => {
    const textBox = screen.getByRole("textbox"),
      searchButton = screen.getByRole("button", {
        name: /search/i,
      }),
      tableBody = screen.getAllByRole("rowgroup")[1],
      user = userEvent.setup();

    expect(textBox).toBeInTheDocument();
    expect(searchButton).toBeInTheDocument();
    expect(textBox).toHaveValue("");

    await user.click(searchButton);

    expect(tableBody).toBeEmptyDOMElement();
  });

  it("it searches for search term using the github api.", async () => {
    const textBox = screen.getByRole("textbox"),
      searchButton = screen.getByRole("button", {
        name: /search/i,
      }),
      user = userEvent.setup();
    // type and search for search term
    await user.type(textBox, searchTerm), await user.click(searchButton);

    expect(textBox).toHaveValue(searchTerm);
    expect(await screen.findByText(searchTerm)).toBeInTheDocument();
  });

  it("navigates from one page to another.", async () => {
    const textBox = screen.getByRole("textbox"),
      searchButton = screen.getByRole("button", {
        name: /search/i,
      }),
      prevButton = screen.getByRole("button", {
        name: "<",
      }),
      nextButton = screen.getByRole("button", {
        name: ">",
      }),
      user = userEvent.setup();

    await user.type(textBox, searchTerm),
      await user.click(searchButton),
      await user.click(nextButton);

    expect(textBox).toHaveValue(searchTerm);
    expect(await screen.findByText(page2Items[0].name)).toBeInTheDocument();
    // go back to previous page
    await user.click(prevButton);
    expect(await screen.findByText(searchTerm)).toBeInTheDocument();
  });
});
