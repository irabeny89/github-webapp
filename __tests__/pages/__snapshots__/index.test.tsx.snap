// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home page test match snapshot 1`] = `
<main
  class="container"
>
  <h1
    class="title"
  >
    GitHub Repositories Search
  </h1>
  <p
    class="description"
  >
    Get started by using the search box...
  </p>
  <div
    class="toast-container position-absolute bottom-0 start-50 translate-middle-x"
  />
  <div
    class="container-fluid"
  >
    <div
      class="justify-content-center row"
    >
      <div
        class="col-sm-10"
      >
        <form
          class=""
        >
          <div
            class="justify-content-center my-3 input-group"
          >
            <div
              class="form-floating"
            >
              <input
                arial-label="search GitHub repositories"
                class="form-control"
                name="search"
                placeholder="Enter your search term"
              />
              <label>
                Enter search term
              </label>
            </div>
            <button
              class="btn btn-primary"
              type="submit"
            >
              Search
            </button>
          </div>
        </form>
      </div>
    </div>
  </div>
  <div
    class="container"
  >
    <div
      class="my-5 row"
    >
      <div
        class="col"
      >
        <div
          class="table-responsive"
        >
          <table
            class="table table-sm table-striped table-bordered table-hover"
          >
            <thead>
              <tr>
                <th>
                  Repo Name
                </th>
                <th>
                  Author Name
                </th>
                <th>
                  Stars
                </th>
              </tr>
            </thead>
            <tbody />
          </table>
        </div>
      </div>
    </div>
  </div>
  <div
    class="container"
  >
    <div
      class="row"
    >
      <div
        class="col"
      >
        <div
          class="btn-group"
          role="group"
        >
          <button
            class="border-3 border-white btn btn-primary"
            type="button"
          >
            &lt;
          </button>
          <button
            class="border-3 border-dark btn btn-outline-secondary"
            type="button"
          >
            &gt;
          </button>
        </div>
      </div>
    </div>
  </div>
  <footer
    class="text-center mt-5"
  >
    Powered by
     
    <span
      class="logo"
    >
      <span
        style="box-sizing: border-box; display: inline-block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: relative; max-width: 100%;"
      >
        <span
          style="box-sizing: border-box; display: block; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; max-width: 100%;"
        >
          <img
            alt=""
            aria-hidden="true"
            src="data:image/svg+xml,%3csvg%20xmlns=%27http://www.w3.org/2000/svg%27%20version=%271.1%27%20width=%2772%27%20height=%2716%27/%3e"
            style="display: block; max-width: 100%; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px;"
          />
        </span>
        <img
          alt="Vercel Logo"
          data-nimg="intrinsic"
          decoding="async"
          src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
          style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
        />
        <noscript />
      </span>
    </span>
    <div>
      Made by Ernest
    </div>
  </footer>
</main>
`;
